src/main/kotlin/app/CacheService.kt

```diff
@@ -0,0 +1,84 @@
+@file:Suppress("SqlResolve")
+
+package app
+
+import app.util.HikariCpDataSource
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import org.slf4j.LoggerFactory
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+
+object CacheService {
+    private val log = LoggerFactory.getLogger(CacheService.javaClass)
+    private val objectMapper = jacksonObjectMapper()
+
+    private fun createTableIfAbsent() {
+        val connection = HikariCpDataSource.connection
+        val statement = connection.createStatement()
+
+        statement.execute(
+            "CREATE TABLE IF NOT EXISTS userinfo (" +
+            "id VARCHAR2 PRIMARY KEY, " +
+            "timestamp TIMESTAMP, " +
+            "data JSON" +
+            ")"
+        )
+    }
+
+    fun selectJsonFromDb(username: String): String? {
+        val connection = HikariCpDataSource.connection
+
+        createTableIfAbsent()
+
+        val preparedStatement = connection.prepareStatement(
+            "SELECT " +
+            "timestamp, " +
+            "data " +
+            "FROM userinfo " +
+            "WHERE id = ?"
+        )
+        preparedStatement.setString(1, username.lowercase())
+
+        val result = preparedStatement.executeQuery()
+        result.use {
+            // guaranteed to be at most one.
+            if (it.next()) {
+                val timestamp = it.getTimestamp(1).toInstant()
+                val diffInHours = ChronoUnit.HOURS.between(timestamp, Instant.now())
+                if (diffInHours <= 6) {
+                    val json: String? = it.getString(2)
```

Comment:
What is `2` ? Can it be accessed by name? If not, maybe we should created `val jsonIndex = 2` ?

Reply:
Great. Replaced to access them by the column name.